editor/src/viewport_tools/tools/spline.rs

```diff
@@ -0,0 +1,270 @@
+			self.fsm_state.update_hints(responses);
+			return;
+		}
+
+		if action == ToolMessage::UpdateCursor {
+			self.fsm_state.update_cursor(responses);
+			return;
+		}
+
+		if let ToolMessage::Spline(SplineMessage::UpdateOptions(action)) = action {
+			match action {
+				SplineOptionsUpdate::LineWeight(line_weight) => self.options.line_weight = line_weight,
+			}
+			return;
+		}
+
+		let new_state = self.fsm_state.transition(action, data.0, data.1, &mut self.data, &self.options, data.2, responses);
+
+		if self.fsm_state != new_state {
+			self.fsm_state = new_state;
+			self.fsm_state.update_hints(responses);
+			self.fsm_state.update_cursor(responses);
+		}
+	}
+
+	fn actions(&self) -> ActionList {
+		use SplineToolFsmState::*;
+
+		match self.fsm_state {
+			Ready => actions!(SplineMessageDiscriminant; Undo, DragStart, DragStop, Confirm, Abort),
+			Drawing => actions!(SplineMessageDiscriminant; DragStop, PointerMove, Confirm, Abort),
+		}
+	}
+}
+
+impl Default for SplineToolFsmState {
+	fn default() -> Self {
+		SplineToolFsmState::Ready
+	}
+}
+#[derive(Clone, Debug, Default)]
+struct SplineToolData {
+	points: Vec<DVec2>,
+	next_point: DVec2,
+	weight: u32,
+	path: Option<Vec<LayerId>>,
+	snap_handler: SnapHandler,
+}
+
+impl Fsm for SplineToolFsmState {
+	type ToolData = SplineToolData;
+	type ToolOptions = SplineOptions;
+
+	fn transition(
+		self,
+		event: ToolMessage,
+		document: &DocumentMessageHandler,
+		tool_data: &DocumentToolData,
+		data: &mut Self::ToolData,
+		tool_options: &Self::ToolOptions,
+		input: &InputPreprocessorMessageHandler,
+		responses: &mut VecDeque<Message>,
+	) -> Self {
+		use SplineMessage::*;
+		use SplineToolFsmState::*;
+
+		let transform = document.graphene_document.root.transform;
+
+		if let ToolMessage::Spline(event) = event {
+			match (self, event) {
+				(Ready, DragStart) => {
+					responses.push_back(DocumentMessage::StartTransaction.into());
+					responses.push_back(DocumentMessage::DeselectAllLayers.into());
+					data.path = Some(document.get_path_for_new_layer());
+
+					data.snap_handler.start_snap(document, document.visible_layers(), true, true);
+					let snapped_position = data.snap_handler.snap_position(responses, input.viewport_bounds.size(), document, input.mouse.position);
+
+					let pos = transform.inverse().transform_point2(snapped_position);
+
+					data.points.push(pos);
+					data.next_point = pos;
+
+					data.weight = tool_options.line_weight;
+
+					responses.push_back(make_operation(data, tool_data, true));
+
+					Drawing
+				}
+				(Drawing, DragStop) => {
+					let snapped_position = data.snap_handler.snap_position(responses, input.viewport_bounds.size(), document, input.mouse.position);
+					let pos = transform.inverse().transform_point2(snapped_position);
+
+					// TODO: introduce comparison threshold when operating with canvas coordinates (https://github.com/GraphiteEditor/Graphite/issues/100)
+					if data.points.last() != Some(&pos) {
```

Comment:
issues #100 is already closed this should probably be adressed here

Reply:
Addressed this TODO for all tools.